source/common/upstream/original_dst_cluster.cc

```diff
@@ -0,0 +1,163 @@
+#include "common/upstream/original_dst_cluster.h"
+
+#include <chrono>
+#include <list>
+#include <string>
+#include <vector>
+
+#include "common/network/address_impl.h"
+#include "common/network/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+OriginalDstCluster::LoadBalancer::LoadBalancer(HostSet& host_set, OriginalDstCluster& parent)
+    : host_set_(host_set), parent_(parent) {
+  // host_set_ is initially empty.
+  host_set_.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                     const std::vector<HostSharedPtr>& hosts_removed) -> void {
+    // Update the hosts map
+    for (auto host : hosts_removed) {
+      ENVOY_LOG(debug, "REMOVING HOST {}", host->address()->asString());
+      auto range = host_map_.equal_range(host->address()->asString());
+      bool erased = false;
+      for (auto it = range.first; it != range.second; ++it) {
+        if (it->second == host) {
+          host_map_.erase(it);
+          erased = true;
+          break; // erase invalidates 'it'
+        }
+      }
+      if (!erased) {
+        ENVOY_LOG(debug, "REMOVING HOST {}: Not Found!", host->address()->asString());
+      }
+    }
+    for (auto host : hosts_added) {
+      ENVOY_LOG(debug, "ADDING HOST {}", host->address()->asString());
+      auto range = host_map_.equal_range(host->address()->asString());
+      bool found = false;
+      for (auto it = range.first; it != range.second; ++it) {
+        if (it->second == host) {
+          ENVOY_LOG(debug, "ADDING HOST {}: EXISTED ALREADY!", host->address()->asString());
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        host_map_.insert({host->address()->asString(), host});
+      }
+    }
+  });
+}
+
+HostConstSharedPtr
+OriginalDstCluster::LoadBalancer::chooseHost(const LoadBalancerContext* context) {
+  if (context) {
+    auto connection = context->downstreamConnection();
+    // Verify that we are not looping back to ourselves!
+    if (connection && connection->usingOriginalDst()) {
+      auto& localAddr = connection->localAddress();
+
+      // Check if a host with the destination address is already in the host set.
+      auto it = host_map_.find(localAddr.asString());
+
+      if (it != host_map_.end()) {
+        // Use the existing host
+        auto host = it->second;
+        ENVOY_LOG(debug, "FOUND HOST {}", host->address()->asString());
+        host->weight(2); // Mark as used.
+        return host;
+      } else {
+        // Add a new host
+        auto localIP = localAddr.ip();
+        if (localIP) {
+          Network::Address::InstanceConstSharedPtr hostIpPort;
```

Comment:
How come we use the local address below as the host? I thought the idea was to use the real address from original_dst? I'm probably missing something major, so apologies if this is a silly question. A comment in the code could also be beneficial to the reader on the rationale here.